File: segment_tree.py
class Node:
    def __init__(self, query, start, end):
        self.start = start
        self.end = end
        self.query = query
        self.left = None
        self.right = None

class Segment:
    def __init__(self, arr):
        self.arr = arr
        self.root = None

    def build_tree(self, arr, start, end):
        if start == end:
            new_node = Node(arr[start], start, end)
            return new_node
        else:
            mid = (start+end)//2
            new_node_ = Node(None, start, end)
            new_node_.left = self.build_tree(arr, start, mid)
            new_node_.right = self.build_tree(arr, mid+1, end)
            return new_node_    

    def build_min_query(self, root, start, end):
        if start == end:
            return root.query
        else:
            mid = (start + end) // 2
            root.query = min(self.build_min_query(root.left,start, mid), self.build_min_query(root.right, mid+1, end))
            return root.query

    def build_max_query(self, root, start, end):
        if start == end:
            return root.query
        else:
            mid = (start + end) // 2
            root.query = max(self.build_max_query(root.left, start, mid), self.build_max_query(root.right, mid+1, end))

    def build_sum_query(self, root, start, end):
        if start == end:
            return root.query
        else:
            mid = (start + end) // 2
            root.query = self.build_sum_query(root.left, start, mid) + self.build_sum_query(root.right, mid+1, end)
            return root.query

    def print_tree(self, root):
        if root:
            print(root.query, root.start, root.end)
            self.print_tree(root.left)
            self.print_tree(root.right)

if __name__ == '__main__':
    n = int(input('Enter the number of elements: '))
    arr = []
    for i in range(n):
        arr.append(int(input('Enter the element: ')))
    print('The array is: ', arr)
    st_tree = Segment(arr)
    st_tree.root = st_tree.build_tree(arr, 0, n-1)
    st_tree.print_tree(st_tree.root)
    print('Enter the operation to build query : ')
    print('1. Minimum')
    print('2. Maximum')
    print('3. Sum')
    n = int(input('Enter operation : '))
    if n == 1:
        st_tree.build_min_query(st_tree.root, 0, len(arr)-1)
    elif n == 2:
        st_tree.build_max_query(st_tree.root, 0, len(arr)-1)
    elif n == 3:
        st_tree.build_sum_query(st_tree.root, 0, len(arr)-1)
    st_tree.print_tree(st_tree.root)

File: Segment tree/SegmentTree.java
import java.util.*;
class Node {
    int start_index, end_index, sum;
    Node left, right;
    Node(int start_index, int end_index, int sum) {
        this.end_index = end_index;
        this.start_index = start_index;
        this.sum = sum;
        this.left = null;
        this.right = null;
    }
}

class SegmentTree {
    private static int[] arr;
    private Node root;

    public SegmentTree(int[] arr) {
        root = build(arr, 0, arr.length - 1);
    }

    private Node build(int[] arr, int start, int end) {
        if (start == end) {
            return new Node(start, end, arr[end]);
        }
        int mid = (start + end)/ 2;
        Node left_node = build(arr, start, mid);
        Node right_node = build(arr, mid+1, end);
        int sum = left_node.sum + right_node.sum;
        Node new_node = new Node(start, end, sum);
        new_node.left = left_node;
        new_node.right = right_node;
        return new_node;
    }

    private int query(int query_start, int query_end) {
        return query(query_start, query_end, root);
//        only a private method can call a private variable and not a public one
    }

    public int query(int query_start, int query_end, Node root) {
        if ((query_start > root.end_index) || (query_end < root.start_index) || (root == null))
            return 0;
        else if (query_start <= root.start_index && query_end >= root.end_index)
            return root.sum;
        int left_sum = query(query_start, query_end, root.left);
        int right_sum = query(query_start, query_end, root.right);
        return left_sum+right_sum;
    }

    public static void main(String[] arg) {
        int n;
        System.out.println("Enter the number of elements : ");
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        arr = new int[n];
        System.out.println("enter elements : ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        SegmentTree st = new SegmentTree(arr);
        System.out.println("enter start and end of query : ");
        int sq = sc.nextInt();
        int se = sc.nextInt();
        System.out.println("the query output : "+st.query(sq, se));
    }
}
